### PR TITLE
tso: fix the TSO reset bug after leader change

### LINKED ISSUE #3423: TSO should not be reset after synchronizing when leader election
### What did you do?
```
[2021/02/19 10:19:41.651 +08:00] [WARN] [wal.go:712] ["slow fdatasync"] [took=4.178056812s] [expected-duration=1s]
[2021/02/19 10:19:41.672 +08:00] [WARN] [lease.go:128] ["lease keep alive failed"] [purpose="pd leader election"] [error="context deadline exceeded"]
[2021/02/19 10:19:41.633 +08:00] [INFO] [lease.go:102] ["lease timeout"] [expire=2021/02/19 10:19:41.722 +08:00] [purpose="pd leader election"]
[2021/02/19 10:19:41.690 +08:00] [WARN] [lease.go:128] ["lease keep alive failed"] [purpose="pd leader election"] [error="context canceled"]
[2021/02/19 10:19:41.712 +08:00] [WARN] [lease.go:128] ["lease keep alive failed"] [purpose="pd leader election"] [error="context canceled"]
[2021/02/19 10:19:41.731 +08:00] [INFO] [lease.go:142] ["stop lease keep alive worker"] [purpose="pd leader election"]
[2021/02/19 10:19:41.808 +08:00] [INFO] [server.go:1258] ["no longer a leader because lease has expired, pd leader will step down"]
[2021/02/19 10:19:41.816 +08:00] [INFO] [coordinator.go:737] ["scheduler has been stopped"] [scheduler-name=label-scheduler] [error="context canceled"]
[2021/02/19 10:19:41.816 +08:00] [INFO] [cluster.go:317] ["metrics are reset"]
[2021/02/19 10:19:41.816 +08:00] [INFO] [cluster.go:319] ["background jobs has been stopped"]
[2021/02/19 10:19:41.816 +08:00] [INFO] [server.go:111] ["region syncer has been stopped"]
[2021/02/19 10:19:41.819 +08:00] [INFO] [coordinator.go:245] ["drive push operator has been stopped"]
[2021/02/19 10:19:41.819 +08:00] [INFO] [cluster.go:338] ["coordinator is stopping"]
[2021/02/19 10:19:41.827 +08:00] [INFO] [coordinator.go:737] ["scheduler has been stopped"] [scheduler-name=balance-hot-region-scheduler] [error="context canceled"]
[2021/02/19 10:19:41.830 +08:00] [WARN] [util.go:144] ["apply request took too long"] [took=115.723365ms] [expected-duration=100ms] [prefix="read-only range "] [request="key:\"/pd/6929484873227404498/gc/safe_point\" "] [response="range_response_count:1 size:76"] []
[2021/02/19 10:19:41.830 +08:00] [INFO] [trace.go:145] ["trace[722025223] range"] [detail="{range_begin:/pd/6929484873227404498/gc/safe_point; range_end:; response_count:1; response_revision:224758; }"] [duration=116.172322ms] [start=2021/02/19 10:19:41.714 +08:00] [end=2021/02/19 10:19:41.830 +08:00] [steps="[\"trace[722025223] 'range keys from in-memory index tree'  (duration: 103.480312ms)\"]"]
[2021/02/19 10:19:42.253 +08:00] [INFO] [coordinator.go:737] ["scheduler has been stopped"] [scheduler-name=balance-region-scheduler] [error="context canceled"]
[2021/02/19 10:19:42.350 +08:00] [INFO] [coordinator.go:737] ["scheduler has been stopped"] [scheduler-name=balance-leader-scheduler] [error="context canceled"]
[2021/02/19 10:19:42.708 +08:00] [INFO] [coordinator.go:107] ["patrol regions has been stopped"]
[2021/02/19 10:19:42.708 +08:00] [INFO] [cluster.go:334] ["coordinator has been stopped"]
[2021/02/19 10:19:42.721 +08:00] [INFO] [tso.go:379] ["reset the timestamp in memory"]
[2021/02/19 10:19:42.802 +08:00] [INFO] [server.go:1183] ["start to campaign pd leader"] [campaign-pd-leader-name=pd-172.16.4.131-2379]
[2021/02/19 10:19:42.806 +08:00] [INFO] [lease.go:60] ["lease granted"] [lease-id=1560065938726156503] [lease-timeout=3] [purpose="pd leader election"]
[2021/02/19 10:19:42.812 +08:00] [INFO] [leadership.go:111] ["check campaign resp"] [resp="{\"header\":{\"cluster_id\":2426450124091047575,\"member_id\":12861506471602918822,\"revision\":224760,\"raft_term\":4},\"succeeded\":true,\"responses\":[{\"Response\":{\"ResponsePut\":{\"header\":{\"revision\":224760}}}}]}"]
[2021/02/19 10:19:42.812 +08:00] [INFO] [leadership.go:120] ["write leaderData to leaderPath ok"] [leaderPath=/pd/6929484873227404498/leader] [purpose="pd leader election"]
[2021/02/19 10:19:42.812 +08:00] [INFO] [server.go:1205] ["campaign pd leader ok"] [campaign-pd-leader-name=pd-172.16.4.131-2379]
[2021/02/19 10:19:42.812 +08:00] [INFO] [server.go:1207] ["setting up the global TSO allocator"]
[2021/02/19 10:19:42.813 +08:00] [INFO] [lease.go:118] ["start lease keep alive worker"] [interval=1s] [purpose="pd leader election"]
**[2021/02/19 10:19:42.816 +08:00] [INFO] [tso.go:188] ["sync and save timestamp"] [last=2021/02/19 10:19:40.469 +08:00] [save=2021/02/19 10:19:45.814 +08:00] [next=2021/02/19 10:19:42.814 +08:00]**
[2021/02/19 10:19:42.818 +08:00] [INFO] [server.go:1298] ["server enable region storage"]
**[2021/02/19 10:19:42.819 +08:00] [INFO] [tso.go:379] ["reset the timestamp in memory"]**
[2021/02/19 10:19:42.826 +08:00] [INFO] [cluster.go:286] ["load stores"] [count=1] [cost=3.196685ms]
[2021/02/19 10:19:42.826 +08:00] [INFO] [cluster.go:297] ["load regions"] [count=46278] [cost=1.456<C2><B5>s]
[2021/02/19 10:19:42.882 +08:00] [INFO] [id.go:121] ["idAllocator allocates a new id"] [alloc-id=98000]
[2021/02/19 10:19:42.882 +08:00] [INFO] [coordinator.go:256] ["coordinator starts to collect cluster information"]
[2021/02/19 10:19:42.886 +08:00] [INFO] [util.go:77] ["load cluster version"] [cluster-version=5.0.0-rc.x]
[2021/02/19 10:19:42.886 +08:00] [WARN] [util.go:79] ["PD version less than cluster version, please upgrade PD"] [PD-version=4.0.0-rc.2-451-g760fb650] [cluster-version=5.0.0-rc.x]
[2021/02/19 10:19:42.886 +08:00] [INFO] [server.go:1249] ["PD cluster leader is ready to serve"] [pd-leader-name=pd-172.16.4.131-2379]
```

PD leader is step down due to some reason and start a new round election. But after synchronizing TSO, it is reset immediately which is not right and the TSO service cannot work anymore.

### What version of PD are you using (`pd-server -V`)?
master

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What problem does this PR solve?

Close #3423.

### What is changed and how it works?

If a PD leader resigns and becomes leader again, due to the old closed `parentCtx`, the Global TSO Allocator will be reset immediately. To fix this bug, we need to update the `parentCtx` every time we set up the Global TSO Allocator.

### Check List

Tests

- Unit test
- Integration test

### CODE REVIEW

**Review by Yisaer — COMMENTED** [review]
LGTM

**Review by rleungx — APPROVED** [review]

**Review by Yisaer — APPROVED** [review]
LGTM

**Comment by JmPotato** [comment]
I used a more reasonable way to fix the bug. PTAL again, thx! @rleungx @Yisaer

File: server/server.go [hunk 1]
```diff
@@ -1204,9 +1206,7 @@ func (s *Server) campaignLeader() {
 	go s.member.KeepLeader(ctx)
 	log.Info("campaign pd leader ok", zap.String("campaign-pd-leader-name", s.Name()))
 
-	log.Info("setting up the global TSO allocator")
-	s.tsoAllocatorManager.SetUpAllocator(ctx, tso.GlobalDCLocation, s.member.GetLeadership())
-	defer s.tsoAllocatorManager.ResetAllocatorGroup(tso.GlobalDCLocation)
+	log.Info("initializing the global TSO allocator")
```
Comment (rleungx) [inline server/server.go:1209]: How about moving it to before line 1215? The rest LGTM

Comment (JmPotato) [reply server/server.go]: Done.

**Pull request merged**